indexing-service/src/test/java/io/druid/indexing/overlord/RemoteTaskRunnerTest.java

```diff
@@ -608,4 +613,69 @@ public void testSortByInsertionTime() throws Exception
     Assert.assertEquals(item2, workItems.get(1));
     Assert.assertEquals(item1, workItems.get(2));
   }
+
+  @Test
+  public void testBlacklistZKWorkers() throws Exception
+  {
+    Period timeoutPeriod = Period.millis(1000);
+    makeWorker();
+    makeRemoteTaskRunner(new TestRemoteTaskRunnerConfig(timeoutPeriod));
+
+    TestRealtimeTask task1 = new TestRealtimeTask(
+            "realtime1",
+            new TaskResource("realtime1", 1),
+            "foo",
+            TaskStatus.success("realtime1"),
+            jsonMapper
+    );
+    Future<TaskStatus> taskFuture1 = remoteTaskRunner.run(task1);
+    Assert.assertTrue(taskAnnounced(task1.getId()));
+    mockWorkerRunningTask(task1);
+    mockWorkerCompleteFailedTask(task1);
+    Assert.assertTrue(taskFuture1.get(TIMEOUT_SECONDS, TimeUnit.SECONDS).isFailure());
+    Assert.assertEquals(0, remoteTaskRunner.getBlackListedWorkers().size());
+    Assert.assertEquals(1,
+            remoteTaskRunner.findWorkerRunningTask(task1.getId()).getCountinouslyFailedTasksCount());
+
+    TestRealtimeTask task2 = new TestRealtimeTask(
+            "realtime2",
+            new TaskResource("realtime2", 1),
+            "foo",
+            TaskStatus.running("realtime2"),
+            jsonMapper
+    );
+    Future<TaskStatus> taskFuture2 = remoteTaskRunner.run(task2);
+    Assert.assertTrue(taskAnnounced(task2.getId()));
+    mockWorkerRunningTask(task2);
+    mockWorkerCompleteFailedTask(task2);
+    Assert.assertTrue(taskFuture2.get(TIMEOUT_SECONDS, TimeUnit.SECONDS).isFailure());
+    Assert.assertEquals(1, remoteTaskRunner.getBlackListedWorkers().size());
+    Assert.assertEquals(2,
+            remoteTaskRunner.findWorkerRunningTask(task2.getId()).getCountinouslyFailedTasksCount());
+
+    remoteTaskRunner.cleanBlackListedNode(remoteTaskRunner.findWorkerRunningTask(task2.getId()));
+
+    // After backOffTime the nodes are whitelisted
+    System.out.println("Sleeping for " + timeoutPeriod.toStandardDuration().getMillis());
+    Thread.sleep(2*timeoutPeriod.toStandardDuration().getMillis());
```

Comment:
minor nit : would be nice if we can avoid sleep, helps in making test more deterministic on slow machines.